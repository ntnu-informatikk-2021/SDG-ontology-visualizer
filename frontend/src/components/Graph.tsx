import React, { useRef, useEffect, useState } from 'react';
import { select, Simulation } from 'd3';
import styles from '../css/Graph.module.css';
import { GraphEdge, GraphNode, Node, Ontology } from '../types/ontologyTypes';
import { getRelations } from '../api/ontologies';
import {
  createForceSimulation,
  drawLinks,
  drawNodeLabels,
  drawNodes,
  resetSimulation,
  updateEdgeLabelPositions,
  updateLabelPositions,
  updateLinkPositions,
  updateNodePositions,
} from '../d3/d3';
import { makePredicateUnique, mapOntologyToGraphEdge, removeDuplicates } from '../common/d3';

const initialNode: Node = {
  prefix: {
    prefix: 'SDG',
    iri: 'http://www.semanticweb.org/aga/ontologies/2017/9/SDG#',
  },
  name: 'Miljø',
  id: 'http://www.semanticweb.org/aga/ontologies/2017/9/SDG#Miljø',
};

const Graph: React.FC = () => {
  const svgref = useRef<SVGSVGElement>(null);
  const [nodes, setNodes] = useState<GraphNode[]>([]);
  const [links, setLinks] = useState<GraphEdge[]>([]);
  const [forceSim, setForceSim] = useState<Simulation<GraphNode, GraphEdge>>(
    (undefined as unknown) as Simulation<GraphNode, GraphEdge>,
  );

  const loadInitialData = async () => {
    const ontologies: Ontology[] = await getRelations(initialNode.id);

    const newNodes: GraphNode[] = ontologies
      .map((ontology) =>
        ontology.Subject.id === initialNode.id ? ontology.Object : ontology.Subject,
      )
      .filter(removeDuplicates);
    newNodes.push(initialNode as GraphNode);

    setNodes(newNodes);

    const newLinks = ontologies.map(mapOntologyToGraphEdge);
    setLinks(newLinks);
  };

  const loadMoreData = async (node: GraphNode) => {
    const ontologies: Ontology[] = await getRelations(node.id);

    const newNodes: GraphNode[] = nodes
      .concat(
        ontologies.map((ontology) =>
          ontology.Subject.id === node.id ? ontology.Object : ontology.Subject,
        ),
      )
      .filter(removeDuplicates);

    const newLinks = links
      .concat(ontologies.map(makePredicateUnique).map(mapOntologyToGraphEdge))
      .filter(removeDuplicates);

    setNodes(newNodes);
    setLinks(newLinks);
  };

  const drawGraph = () => {
    if (!nodes || !links || nodes.length === 0 || links.length === 0) return;

    const svg = select(svgref.current);

    drawLinks(svg, links, '.link', 'none', '#a03', 1);
    drawNodes(svg, nodes, '.node', 10, '#22a', '#22e', loadMoreData);
    drawNodeLabels(svg, nodes, '.nodeLabel');
    // drawEdgeLabels(svg, links, '.edgeLabel');

    // TODO: Simplify this
    if (forceSim) {
      forceSim.on('tick', () => {
        updateLinkPositions(svg, links, '.link');
        updateNodePositions(svg, nodes, '.node');
        updateLabelPositions(svg, nodes, '.nodeLabel');
        updateEdgeLabelPositions(svg, links, '.edgeLabel');
      });
      resetSimulation(forceSim, nodes, links);
    } else {
      const newForceSim = createForceSimulation(nodes, links).on('tick', () => {
        updateLinkPositions(svg, links, '.link');
        updateNodePositions(svg, nodes, '.node');
        updateLabelPositions(svg, nodes, '.nodeLabel');
        updateEdgeLabelPositions(svg, links, '.edgeLabel');
      });
      setForceSim(newForceSim);
    }
  };

  useEffect(() => {
    loadInitialData();
  }, []);

  useEffect(() => {
    drawGraph();
  }, [nodes, links]);

  return (
    <div>
      <svg className={styles.graph} ref={svgref} />
    </div>
  );
};

export default Graph;
