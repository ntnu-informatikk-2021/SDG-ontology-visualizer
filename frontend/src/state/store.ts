import { createStore } from 'redux';
import apiErrorReducer from './reducers/apiErrorReducer';

const store = createStore(apiErrorReducer);

export default store;
