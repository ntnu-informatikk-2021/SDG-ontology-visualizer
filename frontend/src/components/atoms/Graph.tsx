import { Center } from '@chakra-ui/react';
import React, { useEffect, useRef, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getRelations } from '../../api/ontologies';
import GraphSimulation from '../../d3/GraphSimulation';
import useWindowDimensions from '../../hooks/useWindowsDimensions';
import { setError } from '../../state/reducers/apiErrorReducer';
import { selectNode } from '../../state/reducers/ontologyReducer';
import { RootState } from '../../state/store';
import { GraphNode } from '../../types/ontologyTypes';

const Graph: React.FC = () => {
  const { height, width } = useWindowDimensions();
  const svgRef = useRef<SVGSVGElement>(null);
  const selectedNode = useSelector((state: RootState) => state.ontology.selectedNode);
  const dispatch = useDispatch();
  const [simulation, setSimulation] = useState<GraphSimulation>();

  const loadData = async () => {
    if (!simulation || !selectedNode) return;

    const ontologies = await getRelations(selectedNode.id);
    simulation.addData(ontologies, selectedNode);
  };

  const onClickNode = (node: GraphNode): void => {
    dispatch(selectNode(node));
  };

  useEffect(() => {
    if (!svgRef || !svgRef.current) return;
    if (!selectedNode) {
      dispatch(setError(new Error('No nodes selected in Graph')));
      return;
    }
    if (!simulation) {
      setSimulation(
        new GraphSimulation(svgRef.current, width, height - 200, selectedNode, onClickNode),
      );
    } else {
      loadData();
    }
  }, [selectedNode, svgRef, simulation]);

  return (
    <Center mx="auto" my="0">
      <svg id="svgGraph" height={height - 200} width={width} ref={svgRef} />
    </Center>
  );
};

export default Graph;
