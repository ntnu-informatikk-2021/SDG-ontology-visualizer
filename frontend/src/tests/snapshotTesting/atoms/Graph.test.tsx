import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Graph from '../../../components/atoms/Graph';
import store from '../../../state/store';

it('renders when there are no items', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Graph nodeFilter={() => true} />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
