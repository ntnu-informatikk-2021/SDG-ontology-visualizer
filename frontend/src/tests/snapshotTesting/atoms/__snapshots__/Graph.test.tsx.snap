// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when there are no items 1`] = `
<div
  className="css-dxhm2"
>
  <svg
    height="100%"
    id="svgGraph"
    width="100%"
  />
  <button
    aria-label="Fullskjerm"
    className="chakra-button css-17ieupc"
    onClick={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      fill="currentColor"
      focusable={false}
      height="40"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="40"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M64 64v128l48-48 48 48 32-32-48-48 48-48H64zm256 0l48 48-48 48 32 32 48-48 48 48V64H320zM64 320v128h128l-48-48 48-48-32-32-48 48-48-48zm288 0l-32 32 48 48-48 48h128V320l-48 48-48-48z"
      />
    </svg>
  </button>
</div>
`;
